### PR TITLE
Add HostPort support

### PR DESCRIPTION
**Pull request by hedibouattour**

### CODE REVIEW

**Review by AloysAugustin — CHANGES_REQUESTED** [review]
Looks good but I think the cleanup part is missing in `DelVppInterface`.

**Review by AloysAugustin — COMMENTED** [review]
Looks good to me with the latest changes, @sknat could you review as well for the interaction with the cnat?

File: calico-vpp-agent/cni/network_vpp.go [hunk 1]
```diff
@@ -193,6 +193,10 @@ func (s *Server) AddVppInterface(podSpec *storage.LocalPodSpec, doHostSideConf b
 	// Select the first 11 characters of the containerID for the host veth.
 	tunTag := podSpec.NetnsName + "-" + podSpec.InterfaceName
 
+	for idx := range podSpec.HostPorts {
+		s.BindHostPort(&podSpec.HostPorts[idx], podSpec.ContainerIps[0].IP)
```
[the remark this answers is not in the document]
Comment (sknat) [reply calico-vpp-agent/cni/network_vpp.go:197]: We should also choose the first `podSpec.ContainerIps` that matches the ip version of `port.HostIP` in order to support dualstack (aka ipv4 + ipv6)

File: calico-vpp-agent/cni/cni_server.go [hunk 2]
```diff
@@ -67,6 +68,15 @@ func NewLocalPodSpecFromAdd(request *pb.AddRequest) (*storage.LocalPodSpec, erro
 		OrchestratorID: request.Workload.Orchestrator,
 		WorkloadID:     request.Workload.Namespace + "/" + request.Workload.Pod,
 		EndpointID:     request.Workload.Endpoint,
+		HostPorts:      make([]storage.HostPortBinding, 0),
+	}
+	for _, port := range request.Workload.Ports {
+		podSpec.HostPorts = append(podSpec.HostPorts, storage.HostPortBinding{
+			HostPort:      port.HostPort,
+			HostIP:        port.HostIp,
```
Comment (sknat) [inline calico-vpp-agent/cni/cni_server.go:76]: Maybe `net.ParseIP` here, to directly store the parsed `net.IP` ?

File: calico-vpp-agent/cni/network_vpp.go [hunk 3]
```diff
@@ -193,6 +193,18 @@ func (s *Server) AddVppInterface(podSpec *storage.LocalPodSpec, doHostSideConf b
 	// Select the first 11 characters of the containerID for the host veth.
 	tunTag := podSpec.NetnsName + "-" + podSpec.InterfaceName
 
+	for idx := range podSpec.HostPorts {
```
Comment (sknat) [inline calico-vpp-agent/cni/network_vpp.go:196]: Here you might exceed the length of the table, and also we want to expose the ports on all the container addresses. 
Maybe you could use a dedicated function returning the first address that matches 
````golang
for _, hostPort := range podSpec.HostPorts {
   for _, containerAddr := range podSpec.ContainerIps {
       if addrFamilyDiffers (containerAddr, hostPort.HostIP) == nil {
          continue
       }
       err = s.BindHostPort(hostPort, containerAddr)
       ...
   } 
}
````
For the error cleanup, we might want to have `s.BindHostPort` return the EntryID and do `hostPorts.EntryID = id` here. That way if one of the call fails, we can call delete on all the previously created entries.

File: calico-vpp-agent/cni/cni_server.go [hunk 4]
```diff
@@ -98,6 +108,38 @@ func NewLocalPodSpecFromDel(request *pb.DelRequest) *storage.LocalPodSpec {
...
+			},
+		},
+		IsRealIP: true,
+		Proto:    types.TCP,
+		LbType:   types.DefaultLB,
+	}
+	s.log.Infof("(add) %s", entry.String())
+	id, err := s.vpp.CnatTranslateAdd(entry)
+	if err != nil {
+		s.log.Errorf("Error re-injecting cnat entry %s : %v", entry.String(), err)
```
Comment (sknat) [inline calico-vpp-agent/cni/cni_server.go:136]: We could `return error.Wrapf(err, "...")`

File: calico-vpp-agent/cni/storage/storage.go [hunk 5]
```diff
@@ -124,6 +124,18 @@ type LocalPodSpec struct {
...
+
+	//HostPort/HostIP
+	HostPortsSize int `struc:"int16,sizeof=HostPorts"`
+	HostPorts     []HostPortBinding
+}
+
+type HostPortBinding struct {
+	HostPort      uint32
+	HostIPSize    int `struc:"int16,sizeof=HostIP"`
+	HostIP        net.IP
```
Comment (sknat) [inline calico-vpp-agent/cni/storage/storage.go:136]: Maybe ```` `struc:"[16]byte"` ```` to avoid having `HostIPSize`

**Comment by AloysAugustin** [comment]
Holding merge until the matching libcalico-go and cni-plugin PRs are merged so that we can refer to the upstream .proto in calico-vpp-agent/proto/Makefile .

**Review by AloysAugustin — APPROVED** [review]

**Review by sknat — APPROVED** [review]

**Pull request merged**